pkg/kratos/service.go

```diff
@@ -120,6 +155,92 @@ func (s *Service) GetLoginFlow(ctx context.Context, id string, cookies []*http.C
 }
 
+func (s *Service) GetRecoveryFlow(ctx context.Context, id string, cookies []*http.Cookie) (*kClient.RecoveryFlow, []*http.Cookie, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.FrontendApi.GetRecoveryFlow")
+	defer span.End()
+
+	flow, resp, err := s.kratos.FrontendApi().
+		GetRecoveryFlow(ctx).
+		Id(id).
+		Cookie(cookiesToString(cookies)).
+		Execute()
+	if err != nil {
+		s.logger.Debugf("full HTTP response: %v", resp)
+		return nil, nil, err
+	}
+
+	return flow, resp.Cookies(), nil
+}
+
+func (s *Service) GetSettingsFlow(ctx context.Context, id string, cookies []*http.Cookie) (*kClient.SettingsFlow, []*http.Cookie, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.FrontendApi.GetSettingsFlow")
+	defer span.End()
+
+	flow, resp, err := s.kratos.FrontendApi().
+		GetSettingsFlow(ctx).
+		Id(id).
+		Cookie(cookiesToString(cookies)).
+		Execute()
+	if err != nil {
+		s.logger.Debugf("full HTTP response: %v", resp)
+		return nil, nil, err
+	}
+
+	return flow, resp.Cookies(), nil
+}
+
+func (s *Service) UpdateRecoveryFlow(
+	ctx context.Context, flow string, body kClient.UpdateRecoveryFlowBody, cookies []*http.Cookie,
+) (*BrowserLocationChangeRequired, []*http.Cookie, error) {
+	ctx, span := s.tracer.Start(ctx, "kratos.FrontendApi.UpdateRecoveryFlow")
+	defer span.End()
+
+	recovery, resp, err := s.kratos.FrontendApi().
+		UpdateRecoveryFlow(ctx).
+		Flow(flow).
+		UpdateRecoveryFlowBody(body).
+		Cookie(cookiesToString(cookies)).
+		Execute()
+
+	// If the recovery code was invalid, kratos returns a 200 response
+	// with a 4060006 error in the rendered ui messages.
+	// If the recovery code was valid, we expect to get a 422 response from kratos.
+	// That is because the user needs to be redirected to self-service settings page.
+	// The sdk forces us to ake the request with an 'application/json' content-type, whereas Kratos
+	// expects the 'Content-Type' and 'Accept' to be 'application/x-www-form-urlencoded'.
+	// This is not a real error, as we still get the URL to which the user needs to be
+	// redirected to.
+
+	if err != nil && resp.StatusCode != 422 {
+		s.logger.Debugf("full HTTP response: %v", resp)
+		return nil, nil, err
+	}
+
+	if resp.StatusCode == 200 {
+		uiMsg := recovery.GetUi()
+
+		for _, message := range uiMsg.GetMessages() {
+			if message.GetId() == InvalidRecoveryCode {
+				return nil, nil, fmt.Errorf("the recovery code is invalid or has already been used")
+			}
+		}
+	}
+
+	redirectResp := new(ErrorBrowserLocationChangeRequired)
+	err = unmarshalByteJson(resp.Body, redirectResp)
+	if err != nil {
+		s.logger.Debugf("Failed to unmarshal JSON: %s", err)
+		return nil, nil, err
+	}
+
+	// We trasform the kratos response to our own custom response here.
+	// The original kratos response contains an 'Error' field, which we remove
+	// because this is not a real error.
+	returnToResp := BrowserLocationChangeRequired{redirectResp.RedirectBrowserTo}
+
+	return &returnToResp, resp.Cookies(), nil
+}
```

Comment:
I am not sure how I feel about this. Every time the user clicks a button ('Resend code` or `Submit`), this endpoint is called and it returns either `{}` or a url to redirect the user to.

The issue is that, every time right after this call the ui seems to call the get flow endpoint. Wouldn't it make sense if this endpoint returned the whole flow object to make 1 api call instead of 2?

Perhaps I am missing something

Reply:
As discussed the redirect is needed to allow returning to settings after recovery is completed